File: Stack/316. Remove Duplicate Letters/Solution.java
class Solution {
    public String removeDuplicateLetters(String s) {
        int n=s.length();
        char[] map=new char[26];
        boolean[] exist=new boolean[26];
        Deque<Character> stack=new ArrayDeque<>();
        for (int i=0; i<n; i++)
            map[s.charAt(i)-'a']++;
        for (int i=0; i<n; i++){
            char c=s.charAt(i);
            if (!exist[c-'a']){
                //stack.peekLast()>c or stack.peekLast()>=c doesn't matter
                while (!stack.isEmpty()&&map[stack.peekLast()-'a']>0&&stack.peekLast()>c)
                    exist[stack.pollLast()-'a']=false;
                exist[c-'a']=true;
                stack.offerLast(c);
            }
            map[c-'a']--;
        }
        StringBuilder builder=new StringBuilder();
        while (!stack.isEmpty())
            builder.append(stack.pollFirst());
        return builder.toString();
    }
}

File: Stack/316. Remove Duplicate Letters/Solution.py
import collections
class Solution(object):
    def removeDuplicateLetters(self, s):
        """
        :type s: str
        :rtype: str
        """
        n=len(s)
        stack=[]
        record=collections.Counter(s)
        for c in s:
            if c not in stack:
                while stack and record[stack[-1]]>0 and stack[-1]>=c:
                    stack.pop()
                stack.append(c)
            record[c]-=1
        return "".join(stack)
